### PR TITLE
[RHOAIENG-4660] Compare runs - Manage runs page

### PR DESCRIPTION
**Pull request by jpuzz0**
Closes: [RHOAIENG-4660](https://issues.redhat.com/browse/RHOAIENG-4660)

## Description
Manage runs page uses selected runs from the Compare runs page initially as pre-selected table rows (runs), and the user has the ability to select up to 10 total (and change what was initially selected if they want). Selections are maintained between pages, cancelling uses what was initially selected by the Compare runs page, and "Update" will update the selections and bring the user back to the Compare runs page.

## Tests
Cypress tests for main interactions

## How to test
1. Navigate to `http://localhost:4010/experiments/jps-fun-world/85c9d729-bf63-441f-af87-d6898e5db5f0/compareRuns/add`
2. Verify the initial Experiment filter is set to the experiment associated with the ID in your URL path
3. Select up to 10 table rows on a single or multiple pages
4. "Update" and verify you're navigated to the Compare runs page with URL search string updated with all run valid IDs
5. Navigate to the same path from step 1
6. "Cancel" and verify you're navigated to the Compare runs page with no "runs" in the search param
7. Navigate to the same path from step 1, except with `?runs=` following with a comma delimited list of run IDs to verify initial rows are checked (assuming those IDs are valid)
8. Verify filter, row links, pagination, sorting works for the Manage runs table
9. Verify selections between pages persist
10. Verify breadcrumbs bring you to the correct locations

image
image

(cc @yannnz)

## Request review criteria:
Self checklist (all need to be checked):
- [x] The developer has manually tested the changes and verified that the changes work
- [x] Commits have been squashed into descriptive, self-contained units of work (e.g. 'WIP' and 'Implements feedback' style messages have been removed)
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has added tests or explained why testing cannot be added (unit or cypress tests for related changes)

If you have UI changes: 
- [x] Included any necessary screenshots or gifs if it was a UI change.
- [x] Included tags to the UX team if it was a UI/UX change (find relevant UX in the [SMEs](https://github.com/opendatahub-io/odh-dashboard/tree/main/docs/smes.md) section).

After the PR is posted & before it merges:
- [ ] The developer has tested their solution on a cluster by using the image produced by the PR to `main`

### CODE REVIEW

File: frontend/src/components/table/TableBase.tsx [hunk 1]
```diff
@@ -222,7 +224,11 @@ const TableBase = <T,>({
   return (
     <>
       {(toolbarContent || showPagination) && (
-        <Toolbar inset={{ default: 'insetNone' }} customChipGroupContent={<></>}>
+        <Toolbar
+          inset={{ default: 'insetNone' }}
+          className="pf-v5-u-w-100"
```
Comment (Gkrumbach07) [inline frontend/src/components/table/TableBase.tsx:229]: why was this needed? `className="pf-v5-u-w-100"`

File: frontend/src/concepts/pipelines/content/compareRuns/CompareRunsContext.tsx [hunk 2]
```diff
@@ -38,32 +38,38 @@ export const CompareRunsContextProvider = conditionalArea<CompareRunsContextProv
     const runIds = searchParams.get(CompareRunsSearchParam.RUNS);
     return runIds ? runIds.split(',') : [];
   }, [searchParams]);
+  const loadedRef = React.useRef(false);
```
Comment (Gkrumbach07) [inline frontend/src/concepts/pipelines/content/compareRuns/CompareRunsContext.tsx:41]: why switch to using a ref here and not the loaded state from useFetchState? It looks like now on initial navigation to the page, the error/empty state is flashed instead of the loading indicator

Comment (Gkrumbach07) [reply frontend/src/concepts/pipelines/content/compareRuns/CompareRunsContext.tsx:41]: i am guessing this is becuase loaded is set before runs is set, causing a few frames to be incorrect

File: frontend/src/pages/pipelines/global/runs/ManageRunsPage.tsx [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+  if (!loaded && !initialLoaded) {
+    return (
+      <Bullseye>
+        <Spinner />
+      </Bullseye>
+    );
+  }
+
+  if (loaded && totalSize === 0 && !tableProps.filter) {
```
Comment (Gkrumbach07) [inline frontend/src/pages/pipelines/global/runs/ManageRunsPage.tsx:81]: can you add a screenshot for this

File: frontend/src/pages/pipelines/global/runs/ManageRunsTable.tsx [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+  filterProps,
+  ...tableProps
+}) => {
+  const navigate = useNavigate();
+  const [isSelectAllChecked, setIsSelectAllChecked] = React.useState(false);
+  const { namespace } = usePipelinesAPI();
+  const pageRunIds = runs.map(({ run_id: runId }) => runId);
+
+  const { tableProps: checkboxTableProps, toggleSelection } = useCheckboxTable(pageRunIds);
+  const selectedIdsRef = React.useRef<string[]>(selectedRunIds);
```
Comment (Gkrumbach07) [inline frontend/src/pages/pipelines/global/runs/ManageRunsTable.tsx:45]: useCheckboxTable manages selections and select all, why are you managing them outside the hook?

File: frontend/src/pages/pipelines/global/runs/ManageRunsTable.tsx [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+            if (isSelectAllChecked) {
+              setIsSelectAllChecked(false);
+            }
+          }
+        }}
+        selectAll={{
+          onSelect: onSelectAll,
+          selected: isSelectAllChecked,
+        }}
```
Comment (Gkrumbach07) [inline frontend/src/pages/pipelines/global/runs/ManageRunsTable.tsx:115]: this overrides checkboxTableProps.selectAll. is this on purpose?

File: frontend/src/pages/pipelines/global/runs/ManageRunsTable.tsx [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+                  onClick={() =>
+                    navigate(
+                      experimentsCompareRunsRoute(namespace, experimentId, selectedIdsRef.current),
+                    )
+                  }
+                  isDisabled={selectedIdsRef.current.length < 1}
+                >
+                  Update
+                </Button>
+              </Tooltip>
```
Comment (Gkrumbach07) [inline frontend/src/pages/pipelines/global/runs/ManageRunsTable.tsx:153]: missing the tooltip when the button is disabled. You can use isAriaDisabled so the tooltip still works in a disabled state.

File: frontend/src/components/table/TableBase.tsx (hunk 1 above)
Comment (jpuzz0) [reply frontend/src/components/table/TableBase.tsx:229]: Without the width of 100% added here, when there are no actions next to the filter input, the pagination for the table gets pushed to the immediate left of the filter input instead of floating to the far right as intended.

File: frontend/src/concepts/pipelines/content/compareRuns/CompareRunsContext.tsx (hunk 2 above)
Comment (jpuzz0) [reply frontend/src/concepts/pipelines/content/compareRuns/CompareRunsContext.tsx:41]: Once the compare runs page loads, that useFetchState's "loaded" state never changes from "true" even when the array of IDs changes. If you want to solve this some other way, let me know.

File: frontend/src/pages/pipelines/global/runs/ManageRunsTable.tsx (hunk 4 above)
Comment (jpuzz0) [reply frontend/src/pages/pipelines/global/runs/ManageRunsTable.tsx:45]: Because the checkbox selections otherwise get reset between page changes.

File: frontend/src/pages/pipelines/global/runs/ManageRunsTable.tsx (hunk 5 above)
Comment (jpuzz0) [reply frontend/src/pages/pipelines/global/runs/ManageRunsTable.tsx:115]: Yes, for the same reason to your comment here: https://github.com/opendatahub-io/odh-dashboard/pull/2655#discussion_r1548285555

File: frontend/src/pages/pipelines/global/runs/ManageRunsPage.tsx (hunk 3 above)
Comment (jpuzz0) [reply frontend/src/pages/pipelines/global/runs/ManageRunsPage.tsx:81]: image

Comment (Gkrumbach07) [reply frontend/src/pages/pipelines/global/runs/ManageRunsPage.tsx:81]: added it to the description

**Comment by Gkrumbach07** [comment]
organizational suggestion: can you put the ManageRunsPage under the compare runs folder instead of the runs folder

File: frontend/src/concepts/pipelines/content/compareRuns/CompareRunsContext.tsx (hunk 2 above)
Comment (Gkrumbach07) [reply frontend/src/concepts/pipelines/content/compareRuns/CompareRunsContext.tsx:41]: thats correct, it wont. thats just how useFetchState works. loaded will be true when the data is not an error and has loaded before. You should be able to react on runs changing instead of having the loaded ref flip.

Comment (jpuzz0) [reply frontend/src/concepts/pipelines/content/compareRuns/CompareRunsContext.tsx:41]: I don't understand what you're suggesting, but your if-statement in the effect added here; https://github.com/opendatahub-io/odh-dashboard/blob/b7576b6a87b63daf8479d5a498595c7b77938092/frontend/src/concepts/pipelines/content/compareRuns/CompareRunsContext.tsx#L64 is checking for a difference in runs already. This effect triggers before the request for new runs occurs when switching from Manage runs to Compare runs. 

This means that we get invalid runs and those run IDs in the URL get stripped before the request even gets made. How would you like to solve this problem?

Comment (jpuzz0) [reply frontend/src/concepts/pipelines/content/compareRuns/CompareRunsContext.tsx:41]: Alternatively, if we don't have Manage runs be a sub-route of Compare runs, the whole context would re-render when visiting Compare runs (coming from Manage runs) and this wouldn't be an issue.

File: frontend/src/pages/pipelines/global/runs/ManageRunsTable.tsx (hunk 4 above)
Comment (Gkrumbach07) [reply frontend/src/pages/pipelines/global/runs/ManageRunsTable.tsx]: i tinkered around. you can just pass a default set of strings to the usecvheckboxtable hook and then set that to selections right away. this means you dont need the custom select all, on toggle, and use ref selections

File: frontend/src/pages/pipelines/global/runs/ManageRunsTable.tsx (hunk 5 above)
Comment (Gkrumbach07) [reply frontend/src/pages/pipelines/global/runs/ManageRunsTable.tsx]: moving convo there then

File: frontend/src/concepts/pipelines/content/compareRuns/CompareRunsContext.tsx (hunk 2 above)
Comment (Gkrumbach07) [reply frontend/src/concepts/pipelines/content/compareRuns/CompareRunsContext.tsx]: lets do that. move the context under each page respectively. really the context will be mostly useful in the first compare runs page as it will have many components relying on the selections

File: frontend/src/pages/pipelines/global/runs/ManageRunsTable.tsx (hunk 4 above)
Comment (jpuzz0) [reply frontend/src/pages/pipelines/global/runs/ManageRunsTable.tsx]: What default set of strings would one pass to useCheckboxTable? Set what to selections right away?

Comment (Gkrumbach07) [reply frontend/src/pages/pipelines/global/runs/ManageRunsTable.tsx]: that table has a prop for the selectedRunIds. and the state inside the checkbox table hook can eb set with an additional optional input param for defaults

Comment (jpuzz0) [reply frontend/src/pages/pipelines/global/runs/ManageRunsTable.tsx]: I think your last sentence is suggesting updating the useCheckboxTable hook with a new param for default IDs set in state, but I'm not seeing how that would result in the same table behavior as the current implementation. Have you tested this and everything behaves the same way?

Comment (Gkrumbach07) [reply frontend/src/pages/pipelines/global/runs/ManageRunsTable.tsx]: i have. you will also need to then use the selections from the checkbox table hook

Comment (jpuzz0) [reply frontend/src/pages/pipelines/global/runs/ManageRunsTable.tsx]: Those "selections" from the checkbox hook are reset between page changes. Maybe you could elaborate.

Comment (jpuzz0) [reply frontend/src/pages/pipelines/global/runs/ManageRunsTable.tsx]: I asked this question to Yan last week: https://www.figma.com/proto/etbN0ccCcZR1qKCpJXQNAg?node-id=1750-107510#751898481. So there's a different expectation here it seems.

Comment (jpuzz0) [reply frontend/src/pages/pipelines/global/runs/ManageRunsTable.tsx]: pagination

Comment (Gkrumbach07) [reply frontend/src/pages/pipelines/global/runs/ManageRunsTable.tsx]: when i tested it on my end the selection seems to work as expected across pagination

File: frontend/src/components/table/useCheckboxTableBase.ts [hunk 7]
```diff
@@ -17,6 +17,7 @@ const useCheckboxTableBase = <T>(
   setSelectedData: React.Dispatch<React.SetStateAction<T[]>>,
   dataMappingHelper: (selectData: T) => string,
   selectAll?: { selected?: boolean; disabled?: boolean },
+  persistSelections?: boolean,
```
Comment (Gkrumbach07) [inline frontend/src/components/table/useCheckboxTableBase.ts:20]: code quality wise we should put these two into a `opts` param so we can avoid putting undefined in when we dont need to set selectAll

**Review by Gkrumbach07 — COMMENTED** [review]
looks good. just one comment then we can look to merge

File: frontend/src/components/table/useCheckboxTableBase.ts (hunk 7 above)
Comment (jpuzz0) [reply frontend/src/components/table/useCheckboxTableBase.ts]: I disagree that this suggestion improves code quality, and this seems more like a personal preference, but have updated anyways.

Comment (Gkrumbach07) [reply frontend/src/components/table/useCheckboxTableBase.ts]: My reasoning is that as function params start to have multiple optional params you can start to have things like this in an extreme case:

```tsx
const someFunc = (p1: string, p2: string, op1?: string, op2?: string, op3?: string, op4?: string) => {...}

someFunc("one", "two", undefined, undefined, undefined, "optional")
```

With one optional param its not a big deal, two sure, but three definitely want to do something about it. The reason I  flagged it is becuase it was talked about at some point prior as something we wanted to shy away from.

Comment (Gkrumbach07) [reply frontend/src/components/table/useCheckboxTableBase.ts]: There was a jira to look to add a linter rule for it, but i dont think it was added as one doesn't exist

https://issues.redhat.com/browse/RHOAIENG-513
